docs/developer/source-versioning-2024-08-30.md

```diff
@@ -0,0 +1,183 @@
+    "source": IdentifierSchema(IdentifierSchemaParams{
+        Elem:        "source",
+        Description: "The source this table is created from.",
+        Required:    true,
+        ForceNew:    true,
+    }),
+    "upstream_name": {
+        Type:        schema.TypeString,
+        Required:    true,
+        ForceNew:    true,
+        Description: "The name of the table in the upstream database.",
+    },
+    "upstream_schema_name": {
+        Type:        schema.TypeString,
+        Optional:    true,
+        ForceNew:    true,
+        Description: "The schema of the table in the upstream database.",
+    },
+    "text_columns": {
+        Description: "Columns to be decoded as text.",
+        Type:        schema.TypeList,
+        Elem:        &schema.Schema{Type: schema.TypeString},
+        Optional:    true,
+        ForceNew:    true,
+    },
+    // ... other fields as needed ...
+}
+```
+
+### 4.2 Resource Implementation
+
+#### 4.2.1 Table From Source Resource
+
+Implement CRUD operations for the new `materialize_table_from_source` resource:
+
+```go
+func TableFromSource() *schema.Resource {
+    return &schema.Resource{
+        CreateContext: tableFromSourceCreate,
+        ReadContext:   tableFromSourceRead,
+        UpdateContext: tableFromSourceUpdate,
+        DeleteContext: tableFromSourceDelete,
+        Importer: &schema.ResourceImporter{
+            StateContext: schema.ImportStatePassthroughContext,
+        },
+        Schema: tableFromSourceSchema,
+    }
+}
+```
+
+The `CreateContext` function will use the new SQL syntax:
+
+```sql
+CREATE TABLE <database_name>.<schema_name>.<name> FROM SOURCE <source_name> (REFERENCE = <upstream name>) WITH (TEXT COLUMNS = (..), ..)
+```
+
+#### 4.2.2 Update Existing Source Resources
+
+Modify the CRUD operations for existing source resources to handle the deprecation of subsource-related fields:
+
+- In `Create` and `Update` operations, if the deprecated `table` field is used, log a warning message advising users to migrate to the new `materialize_table_from_source` resource.
+- In `Read` operations, continue to populate the `table` field if it exists in the state, but also log a deprecation warning.
+
+### 4.3 Migration Strategy
+
+We will not create separate resources with a v2 suffix for sources. Instead, we'll use a gradual migration approach:
+
+1. Deprecate the `table` field in existing source resources.
+2. Introduce the new `materialize_table_from_source` resource.
+3. Allow both old and new configurations to coexist during a transition period.
+
+This approach allows users to migrate their configurations gradually:
+
+- Existing sources can still be created and managed.
+- New tables (formerly subsources) will be created as separate `materialize_table_from_source` resources.
+- Users can migrate their configurations at their own pace by replacing `table` blocks with `materialize_table_from_source` resources.
```

Comment:
This is really smart! Will you be able to mix-n-match `table` and `materialize_source_table` resources for the same source? Or will you always have to migrate an entire source at once?

My hunch is that we should require that you only use one or the other on a per source basis. That seems likely to make the code much simpler and easier to reason about. And it still affords the user quite a bit of flexibility to do the migration incrementally by migrating one source at a time.

Reply:
>  Will you be able to mix-n-match table and materialize_source_table resources for the same source? Or will you always have to migrate an entire source at once?

Based on what I've tested so far using main, it looks like that you could have both running simultaneously. So if we were to take the following example:

- Existing standard Postgres source:

```hcl
resource "materialize_source_postgres" "pg_source" {
  name         = "pg_source"
  cluster_name = "quickstart"

  expose_progress {
    name = "expose_postgres_progress"
  }

  postgres_connection {
    name          = materialize_connection_postgres.postgres_connection.name
  }
  publication = "mz_source"
  table {
    upstream_name  = "table1"
    upstream_schema_name = "public"
  }
  table {
    upstream_name  = "table2"
    upstream_schema_name = "public"
  }
  ...
}
```

- Then define a new `materialize_source_table`:

```hcl
resource "materialize_source_table" "source_table3" {
  name           = "source_table3"
  schema_name    = "public"
  database_name  = "materialize"

  source {
    name          = materialize_source_postgres.pg_source.name
    schema_name   = "public"
    database_name = "materialize"
  }

  upstream_name         = "table2"
  upstream_schema_name  = "public"

  text_columns = [
    "updated_at"
  ]
}
```

So the user should still be able to slowly do the migration if needed as the `table` blocks and the `materialize_source_table` are independent, at least until we do the actual migration on the Materialize side, then in that case, users would just need to define all of their tables as `materialize_source_table` and as long as we have all of the required information in the catalog tables, we should be able to update the state accordingly.

Reply:
>  it looks like that you could have both running simultaneously

Yes at the moment we allow both if you have the new `enable_create_table_from_source` feature-flag enabled, however we could artificially enforce that you are using either the old model or the new model in the terraform provider on a per-source basis to make things simpler?

Reply:
Ah ok, if it's going to be supported to have both simultaneously on the database side, then maybe it's actually simpler to exactly mirror that in Terraform?

Reply:
Yep, it would be simpler if we could have it as it is at the moment. Allowing us to have both simultaneously on the database side makes things easier for the Terraform implementation as well.